p2p/addressbook/addrbook.go

```diff
@@ -283,46 +228,113 @@ func (a *addrBook) pickTried(bucket int) *knownAddress {
 }
 
 // NumAddresses returns the number of addresses known to the address manager.
-func (a *addrBook) numAddresses() int {
+func (a *AddrBook) numAddresses() int {
 	return a.nTried + a.nNew
 }
 
 // NumAddresses returns the number of addresses known to the address manager.
-func (a *addrBook) NumAddresses() int {
+func (a *AddrBook) NumAddresses() int {
 	a.mu.Lock()
 	defer a.mu.Unlock()
 	return a.numAddresses()
 }
 
+// getAddressFromBuckets selects an address from given slice of buckets.
+func (a *AddrBook) getAddressFromBuckets(buckets []map[peer.ID]*knownAddress) *knownAddress {
+	nonEmptyBuckets := make([]int, 0, len(buckets)) // get non-empty buckets ids for reduce extracting time.
+	for i := range buckets {
+		if len(buckets[i]) > 0 {
+			nonEmptyBuckets = append(nonEmptyBuckets, i)
+		}
+	}
+	if len(nonEmptyBuckets) == 0 {
+		return nil // no addresses to extract.
+	}
+	large := 1 << 30
+	factor := 1.0
+	for {
+		// pick a random bucket.
+		bucket := nonEmptyBuckets[a.rand.Intn(len(nonEmptyBuckets))]
+		if len(buckets[bucket]) == 0 {
+			continue
+		}
+
+		// Pick a random entry in it.
+		var ka *knownAddress
+		nth := a.rand.Intn(len(buckets[bucket]))
+		for _, value := range buckets[bucket] {
+			if nth == 0 {
+				ka = value
+			}
+			nth--
+		}
```

Comment:
why ka is not buckets[bucket][nth] ? maybe i just dont understand this loop

Reply:
structure is `map[peer.ID]*knownAddress`, so we should iterate in loop, cause rand return only int